File: it-agents/skills/code-architect/references/code-examples/java/Adapter.java
// Adapter — make an incompatible interface usable via a wrapper.

// Target interface the client expects.
interface DataSource {
    String read();
}

// Adaptee with an incompatible API.
class LegacyFeed {
    byte[] fetchBytes() { return new byte[] { 72, 105 }; } // "Hi"
}

// Adapter conforms to DataSource while delegating to the adaptee.
class LegacyFeedAdapter implements DataSource {
    private final LegacyFeed feed;
    LegacyFeedAdapter(LegacyFeed feed) { this.feed = feed; }
    public String read() {
        return new String(feed.fetchBytes());
    }
}

public class Adapter {
    static void consume(DataSource src) {
        System.out.println("client got: " + src.read());
    }

    public static void main(String[] args) {
        DataSource adapted = new LegacyFeedAdapter(new LegacyFeed());
        consume(adapted);
    }
}

File: it-agents/skills/code-architect/references/code-examples/swift/Adapter.swift
// Adapter — make an incompatible interface usable via a wrapper.

// Target interface the client expects.
protocol DataSource {
    func fetchRows() -> [String]
}

// Incompatible existing service with a different shape.
final class LegacyFeed {
    func dumpPayload() -> String {
        "alpha;beta;gamma"
    }
}

// Adapter conforms to the target while delegating to the legacy service.
struct LegacyFeedAdapter: DataSource {
    private let legacy: LegacyFeed

    init(_ legacy: LegacyFeed) {
        self.legacy = legacy
    }

    func fetchRows() -> [String] {
        legacy.dumpPayload().split(separator: ";").map(String.init)
    }
}

func printRows(from source: DataSource) {
    for (index, row) in source.fetchRows().enumerated() {
        print("row \(index): \(row)")
    }
}

func runDemo() {
    let adapted = LegacyFeedAdapter(LegacyFeed())
    printRows(from: adapted)
}

runDemo()
